components/net/http_cache.rs

```diff
@@ -0,0 +1,486 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#![deny(missing_docs)]
+
+//! A memory cache Implements the logic specified in http://tools.ietf.org/html/rfc7234
+//! and <http://tools.ietf.org/html/rfc7232>.
+
+use fetch::methods::DoneChannel;
+use http_loader::is_redirect_status;
+use hyper::header;
+use hyper::header::ContentType;
+use hyper::header::Headers;
+use hyper::method::Method;
+use hyper::status::StatusCode;
+use hyper_serde::Serde;
+use net_traits::{Metadata, FetchMetadata, ReferrerPolicy};
+use net_traits::request::Request;
+use net_traits::response::{HttpsState, Response, ResponseBody};
+use servo_url::ServoUrl;
+use std::collections::HashMap;
+use std::str;
+use std::sync::{Arc, Mutex};
+use time;
+use time::{Duration, Tm};
+use url::Url;
+
+
+
+/// The key used to differentiate requests in the cache.
+#[derive(Clone, Eq, Hash, PartialEq)]
+pub struct CacheKey {
+    url: ServoUrl
+}
+
+impl CacheKey {
+    fn new(request: Request) -> CacheKey {
+        CacheKey {
+            url: request.url().clone()
+        }
+    }
+
+    fn from_url_string(url: &str) -> Result<CacheKey, ()> {
+        if let Ok(url) = Url::parse(&url) {
+            if !url.has_host() {
```

Comment:
If I understand correctly, this is supposed to address my question about `Location: foo.html`? As my response (hopefully) explained, I don't believe this is the correct solution; this code would not be reached, because Url::parse would return an error for a relative URL.

Reply:
Ok I see, would it make sense if I were to use the concat the request origin with the value of the location header as a way to find the relevant entry by url?

Reply:
Use `request.current_url().join(url)`; that uses the request's current URL as the base and parses the string relative to it. A non-relative URL will override it, too.

Reply:
thanks